woocommerce/woocommerce-gutenberg-products-block
# Revise checkout payment statuses to avoid data loss on error

We're seeing errors and inconsistencies with payment method data being lost when errors cause a status change (to pristine) on checkout. On reviewing this, we've also noticed some discrepancies with how statuses (in particular the `STARTED` status) are used in a non-documented way. It's also confusing how some data is stored in the context provider, some in hooks, and yet it all needs to be kept in sync (currently with useEffect calls).

To solve this, I've taken the following approach:

- Existing payment statuses are left in place ([unlike this PR which added an IDLE status](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5318)). The main difference here is, like specified in the docs, STARTED is reserved only for Express Payment Methods. Before this PR, saved methods were also using this status to add extra data (tokens).
- Fixed PRISTINE status so that it only removes errors, not payment method data.
- Removed the `useActivePaymentMethod` hook which had methods to set active method, and active token. This has been replaced with data in the context provider.
- `setActivePaymentMethod` now accepts payment method data. By consolidating it within this function, saved methods no longer need to use the `STARTED` status to store data.
- Consolidated active payment method with active token. The active token can be derived from the active payment method's defined data. Removed `setActiveToken`.
- When a different method is made active, all payment data is cleared so that it doesn't linger for other methods.
- The logic for selecting a default gateway was handled in two places. Saved methods, and the main context provider. I've combined this into one `useEffect` hook to avoid conflicts.

Fixes #5296
Closes #5318 in favor of this.

### Testing

Please smoke test the following:

1. Payment using a core gateway (BACS, Check)
2. Payment via Stripe (using a new test card)
3. Payment via a saved card (with Stripe)
4. Payment via an express payment method (also Stripe)
5. Start an express payment and cancel it. Confirm the previously selected gateway is selected after cancellation.

Also, test the case from #5296.

1. Go to checkout and select a saved method for payment
2. Leave a required field blank and place order
3. See error
4. Correct mistake and place order again. Order should go through.
 
### Changelog

> Fixed a case where payments could fail after validation errors when using saved cards.
